src/Confluent.Kafka/Producer.cs

```diff
@@ -151,96 +269,192 @@ private void setAndValidateSerializers(Serializer<TKey> keySerializer, Serialize
+            if (deliveryReportEnabledFieldsObj != null)
+            {
+                var fields = deliveryReportEnabledFieldsObj.ToString().Replace(" ", "");
+                if (fields != "all")
+                {
+                    this.enableDeliveryReportKey = false;
+                    this.enableDeliveryReportValue = false;
+                    this.enableDeliveryReportHeaders = false;
+                    this.enableDeliveryReportTimestamp = false;
+                    if (fields != "none")
+                    {
+                        var parts = fields.Split(',');
+                        foreach (var part in parts)
+                        {
+                            switch (part)
+                            {
+                                case "key": this.enableDeliveryReportKey = true; break;
+                                case "value": this.enableDeliveryReportValue = true; break;
+                                case "timestamp": this.enableDeliveryReportTimestamp = true; break;
+                                case "headers": this.enableDeliveryReportHeaders = true; break;
+                                default: throw new ArgumentException(
+                                    $"Unexpected delivery report field name '{part}' in config value '{ConfigPropertyNames.ProducerDeliveryReportFields}'.");
+                            }
+                        }
+                    }
+                }
+            }
+
+            // Note: changing the default value of produce.offset.report at the binding level is less than
+            // ideal since it means the librdkafka configuration docs will no longer completely match the 
+            // .NET client. The default should probably be changed in librdkafka as well.
+            if (modifiedConfig.FirstOrDefault(prop => prop.Key == "produce.offset.report").Value == null)
+            {
+                modifiedConfig = modifiedConfig.Concat(new KeyValuePair<string, string>[] { new KeyValuePair<string, string>("produce.offset.report", "true") });
+            }
+
+            var configHandle = SafeConfigHandle.Create();
+
+            modifiedConfig.ToList().ForEach((kvp) => {
+                if (kvp.Value == null) throw new ArgumentException($"'{kvp.Key}' configuration parameter must not be null.");
+                configHandle.Set(kvp.Key, kvp.Value.ToString());
+            });
+
+
+            IntPtr configPtr = configHandle.DangerousGetHandle();
+
+            if (enableDeliveryReports)
+            {
+                Librdkafka.conf_set_dr_msg_cb(configPtr, DeliveryReportCallback);
+            }
+
+            // Explicitly keep references to delegates so they are not reclaimed by the GC.
+            errorCallbackDelegate = ErrorCallback;
+            logCallbackDelegate = LogCallback;
+            statsCallbackDelegate = StatsCallback;
+
+            // TODO: provide some mechanism whereby calls to the error and log callbacks are cached until
```

Comment:
ahh yes, I forgot about this. I want to exclude it from the scope of this PR because I want to get an API close to the final one in a release ASAP. have noted.